File: Documentation/5.4/Samples/csharp/Raven.Documentation.Samples/Indexes/IndexingRelatedDocuments.cs
using System.Collections.Generic;
using System.Linq;
using Raven.Client.Documents;
using Raven.Client.Documents.Indexes;
using Raven.Documentation.Samples.Indexes.IndexingRelatedDocuments;
using Raven.Documentation.Samples.Orders;
using Raven.Client.Documents.Linq;

namespace Raven.Documentation.Samples.Indexes
{
    namespace Syntax
    {
        public class LoadDocumentSyntax
        {
            private interface ILoadDocument
            {
                #region syntax
                T LoadDocument<T>(string relatedDocumentId);
                
                T LoadDocument<T>(string relatedDocumentId, string relatedCollectionName);

                T[] LoadDocument<T>(IEnumerable<string> relatedDocumentIds);

                T[] LoadDocument<T>(IEnumerable<string> relatedDocumentIds, string relatedCollectionName);
                #endregion
            }
        }
    }
    
    namespace IndexingRelatedDocuments
    {
        #region indexing_related_documents_1
        public class Products_ByCategoryName : AbstractIndexCreationTask<Product>
        {
            public class IndexEntry
            {
                public string CategoryName { get; set; }
            }

            public Products_ByCategoryName()
            {
                Map = products => from product in products
                    
                    // Call LoadDocument to load the related Category document
                    // The document ID to load is specified by 'product.Category'
                    let category = LoadDocument<Category>(product.Category)
                    
                    select new IndexEntry
                    {
                        // Index the Name field from the related Category document
                        CategoryName = category.Name
                    };
                        
                    // Since NoTracking was Not specified,
                    // then any change to either Products or Categories will trigger reindexing 
            }
        }
        #endregion
        
        #region indexing_related_documents_1_JS
        public class Products_ByCategoryName_JS : AbstractJavaScriptIndexCreationTask
        {
            public Products_ByCategoryName_JS()
            {
                Maps = new HashSet<string>()
                {
                    // Call method 'load' to load the related Category document
                    // The document ID to load is specified by 'product.Category'
                    // The Name field from the related Category document will be indexed
                    
                    @"map('products', function(product) {
                        let category = load(product.Category, 'Categories')
                        return {
                            CategoryName: category.Name
                        };
                    })"
                    
                    // Since noTracking was Not specified,
                    // then any change to either Products or Categories will trigger reindexing 
                };
            }
        }
        #endregion
        
        #region indexing_related_documents_3
        // The referencing document
        public class Author
        {
            public string Id { get; set; }
            public string Name { get; set; }
            
            // Referencing a list of related document IDs
            public List<string> BookIds { get; set; }
        }
        
        // The related document
        public class Book
        {
            public string Id { get; set; }
            public string Name { get; set; }
        }

        #endregion

        #region indexing_related_documents_4
        public class Authors_ByBooks : AbstractIndexCreationTask<Author>
        {
            public class IndexEntry
            {
                public IEnumerable<string> BookNames { get; set; }
            }

            public Authors_ByBooks()
            {
                Map = authors => from author in authors
                    select new IndexEntry
                    {
                        // For each Book ID, call LoadDocument and index the book's name
                        BookNames = author.BookIds.Select(x => LoadDocument<Book>(x).Name)
                    };
                
                // Since NoTracking was Not specified,
                // then any change to either Authors or Books will trigger reindexing 
            }
        }
        #endregion
        
        #region indexing_related_documents_4_JS
        public class Authors_ByBooks_JS : AbstractJavaScriptIndexCreationTask
        {
            public Authors_ByBooks_JS()
            {
                Maps = new HashSet<string>()
                {
                    // For each Book ID, call 'load' and index the book's name
                    @"map('Author', function(author) {
                        return {
                            Books: author.BooksIds.map(x => load(x, 'Books').Name)
                        }
                    })"
                    
                    // Since NoTracking was Not specified,
                    // then any change to either Authors or Books will trigger reindexing 
                };
            }
        }
        #endregion
        
        #region indexing_related_documents_6
        public class Products_ByCategoryName_NoTracking : AbstractIndexCreationTask<Product>
        {
            public class IndexEntry
            {
                public string CategoryName { get; set; }
            }

            public Products_ByCategoryName_NoTracking()
            {
                Map = products => from product in products
                    
                    // Call NoTracking.LoadDocument to load the related Category document w/o tracking
                    let category = NoTracking.LoadDocument<Category>(product.Category)
                    
                    select new IndexEntry
                    {
                        // Index the Name field from the related Category document
                        CategoryName = category.Name
                    };
                        
                    // Since NoTracking is used -
                    // then only the changes to Products will trigger reindexing 
            }
        }
        #endregion
        
        #region indexing_related_documents_6_JS
        public class Products_ByCategoryName_NoTracking_JS : AbstractJavaScriptIndexCreationTask
        {
            public Products_ByCategoryName_NoTracking_JS()
            {
                Maps = new HashSet<string>()
                {
                    // Call 'noTracking.load' to load the related Category document w/o tracking
                    
                    @"map('products', function(product) {
                        let category = noTracking.load(product.Category, 'Categories')
                        return {
                            CategoryName: category.Name
                        };
                    })"
                    
                    // Since noTracking is used -
                    // then only the changes to Products will trigger reindexing
                };
            }
        }
        #endregion
    }

    public class IndexingRelatedDocumentsQueries
    {
        public async void Queries()
        {
            using (var store = new DocumentStore())
            {
                using (var session = store.OpenSession())
                {
                    #region indexing_related_documents_2
                    IList<Product> matchingProducts = session
                        .Query<Products_ByCategoryName.IndexEntry, Products_ByCategoryName>()
                        .Where(x => x.CategoryName == "Beverages")
                        .OfType<Product>()
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region indexing_related_documents_2_async
                    IList<Product> matchingProducts = await asyncSession
                        .Query<Products_ByCategoryName.IndexEntry, Products_ByCategoryName>()
                        .Where(x => x.CategoryName == "Beverages")
                        .OfType<Product>()
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region indexing_related_documents_5
                    // Get all authors that have books with title: "The Witcher"
                    IList<Author> matchingAuthors = session
                        .Query<Authors_ByBooks.IndexEntry, Authors_ByBooks>()
                        .Where(x => x.BookNames.Contains("The Witcher"))
                        .OfType<Author>()
                        .ToList();
                    #endregion
                }
                
                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region indexing_related_documents_5_async
                    // Get all authors that have books with title: "The Witcher"
                    IList<Author> matchingAuthors = await asyncSession
                        .Query<Authors_ByBooks.IndexEntry, Authors_ByBooks>()
                        .Where(x => x.BookNames.Contains("The Witcher"))
                        .OfType<Author>()
                        .ToListAsync();
                    #endregion
                }
                
                using (var session = store.OpenSession())
                {
                    #region indexing_related_documents_7
                    IList<Product> matchingProducts = session
                        .Query<Products_ByCategoryName_NoTracking.IndexEntry, Products_ByCategoryName_NoTracking>()
                        .Where(x => x.CategoryName == "Beverages")
                        .OfType<Product>()
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region indexing_related_documents_7_async
                    IList<Product> matchingProducts = await asyncSession
                        .Query<Products_ByCategoryName_NoTracking.IndexEntry, Products_ByCategoryName_NoTracking>()
                        .Where(x => x.CategoryName == "Beverages")
                        .OfType<Product>()
                        .ToListAsync();
                    #endregion
                }
            }
        }
    }
}

File: Documentation/5.4/Samples/php/Indexes/IndexingRelatedDocuments.php
<?php

namespace RavenDB\Samples\Indexes;

use Product;
use RavenDB\Documents\DocumentStore;
use RavenDB\Documents\Indexes\AbstractIndexCreationTask;
use RavenDB\Documents\Indexes\AbstractJavaScriptIndexCreationTask;
use RavenDB\Type\StringArray;

class IndexingRelatedDocuments
{
    public function queries(): void
    {
        $store = new DocumentStore();
        try {
            $session = $store->openSession();
            try {
                # region indexing_related_documents_2
                $matchingProducts = $session
                    ->query(Products_ByCategoryName_IndexEntry::class, Products_ByCategoryName::class)
                    ->whereEquals("CategoryName", "Beverages")
                    ->ofType(Product::class)
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region indexing_related_documents_5
                // Get all authors that have books with title: "The Witcher"
                $matchingAuthors = $session
                    ->query(Authors_ByBooks_IndexEntry::class, Authors_ByBooks::class)
                    ->containsAny("BookNames", ["The Witcher"])
                    ->ofType(Author::class)
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region indexing_related_documents_7
                $matchingProducts = $session
                    ->query(Products_ByCategoryName_NoTracking_IndexEntry::class, Products_ByCategoryName_NoTracking::class)
                    ->whereEquals("CategoryName", "Beverages")
                    ->ofType(Product::class)
                    ->toList();

                # endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}

/*
interface ILoadDocument
{
    # region syntax
    T LoadDocument<T>(string relatedDocumentId);

    T LoadDocument<T>(string relatedDocumentId, string relatedCollectionName);

    T[] LoadDocument<T>(IEnumerable<string> relatedDocumentIds);

    T[] LoadDocument<T>(IEnumerable<string> relatedDocumentIds, string relatedCollectionName);
    # endregion
}
*/

# region indexing_related_documents_1
class Products_ByCategoryName_IndexEntry
{
    private ?string $categoryName = null;

    public function getCategoryName(): ?string
    {
        return $this->categoryName;
    }

    public function setCategoryName(?string $categoryName): void
    {
        $this->categoryName = $categoryName;
    }
}
class Products_ByCategoryName extends AbstractIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->map =
            "from product in docs.Products " .
            'let category = this.LoadDocument(product.Category, "Categories") ' .
            "select new { CategoryName = category.Name }";

            // Since NoTracking was Not specified,
            // then any change to either Products or Categories will trigger reindexing
    }
}
# endregion

# region indexing_related_documents_1_JS
class Products_ByCategoryName_JS extends AbstractJavaScriptIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        // Call method 'load' to load the related Category document
        // The document ID to load is specified by 'product.Category'
        // The Name field from the related Category document will be indexed
        $this->setMaps([
            "map('products', function(product) { " .
            "    let category = load(product.Category, 'Categories') " .
            "    return { " .
            "        CategoryName: category.Name " .
            "    }; " .
            "})"
        ]);

        // Since noTracking was Not specified,
        // then any change to either Products or Categories will trigger reindexing

    }
}
# endregion

# region indexing_related_documents_3
// The referencing document
class Author
{
    private ?string $id = null;
    private ?string $name = null;

    // Referencing a list of related document IDs
    private ?StringArray $bookIds = null;

    public function getId(): ?string
    {
        return $this->id;
    }

    public function setId(?string $id): void
    {
        $this->id = $id;
    }

    public function getName(): ?string
    {
        return $this->name;
    }

    public function setName(?string $name): void
    {
        $this->name = $name;
    }

    public function getBookIds(): ?StringArray
    {
        return $this->bookIds;
    }

    public function setBookIds(?StringArray $bookIds): void
    {
        $this->bookIds = $bookIds;
    }
}

// The related document
class Book
{
    private ?string $id = null;
    private ?string $name = null;

    public function getId(): ?string
    {
        return $this->id;
    }

    public function setId(?string $id): void
    {
        $this->id = $id;
    }

    public function getName(): ?string
    {
        return $this->name;
    }

    public function setName(?string $name): void
    {
        $this->name = $name;
    }
}
# endregion

# region indexing_related_documents_4
class Authors_ByBooks_IndexEntry
{
    private ?StringArray $bookNames = null;

    public function getBookNames(): ?StringArray
    {
        return $this->bookNames;
    }

    public function setBookNames(?StringArray $bookNames): void
    {
        $this->bookNames = $bookNames;
    }
}
class Authors_ByBooks extends AbstractIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->map =
            "from author in docs.Authors " .
            "select new " .
            "{" .
            // For each Book ID, call LoadDocument and index the book's name
            '    BookNames = author.BookIds.Select(x => LoadDocument(x, "Books").Name)' .
            "}";

        // Since NoTracking was Not specified,
        // then any change to either Authors or Books will trigger reindexing
    }
}
# endregion

# region indexing_related_documents_4_JS
class Authors_ByBooks_JS extends AbstractJavaScriptIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->setMaps([
            // For each Book ID, call 'load' and index the book's name
            "map('Author', function(author) {
                return {
                    Books: author.BooksIds.map(x => load(x, 'Books').Name)
                }
            })"
        ]);

        // Since NoTracking was Not specified,
        // then any change to either Authors or Books will trigger reindexing
    }
}
# endregion

# region indexing_related_documents_6
class Products_ByCategoryName_NoTracking_IndexEntry
{
    private ?string $categoryName = null;

    public function getCategoryName(): ?string
    {
        return $this->categoryName;
    }

    public function setCategoryName(?string $categoryName): void
    {
        $this->categoryName = $categoryName;
    }
}

class Products_ByCategoryName_NoTracking extends AbstractIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->map =
            "from product in docs.Products " .
            # Call NoTracking.LoadDocument to load the related Category document w/o tracking
            'let category = NoTracking.LoadDocument(product.Category, "Categories") ' .
            "select new {" .
            # Index the name field from the related Category document
            " CategoryName = category.Name " .
            "}";

        // Since NoTracking is used -
        // then only the changes to Products will trigger reindexing
    }
}
# endregion

# region indexing_related_documents_6_JS
class Products_ByCategoryName_NoTracking_JS extends AbstractJavaScriptIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->setMaps([
            // Call 'noTracking.load' to load the related Category document w/o tracking
            "map('products', function(product) {
                let category = noTracking.load(product.Category, 'Categories')
                return {
                    CategoryName: category.Name
                };
            })"
        ]);

        // Since noTracking is used -
        // then only the changes to Products will trigger reindexing
    }
}
# endregion
